zcash_client_backend/src/zip321.rs

```diff
@@ -0,0 +1,911 @@
+                    match payment.recipient_address {
+                        Address::Sapling(_) => payment.memo = Some(m),
+                        Address::Transparent(_) => return Err(format!("Payment {} attempted to associate a memo with a transparent recipient address", i)),
+                    }
+                },
+
+                Param::Label(m) => payment.label = Some(m),
+                Param::Message(m) => payment.message = Some(m),
+                Param::Other(n, v) => payment.other_params.push((n, v)),
+                _otherwise => {}
+            }
+        }
+
+        return Ok(payment);
+    }
+
+    /// Parser that consumes the leading "zcash:[address]" from
+    /// a zip-0321 URI.
+    pub fn lead_addr<'a, P: consensus::Parameters>(
+        params: &'a P,
+    ) -> impl Fn(&str) -> IResult<&str, Option<IndexedParam>> + 'a {
+        move |input: &str| {
+            let (input, _) = tag("zcash:")(input)?;
+            let (input, addr) = map_res(take_until("?"), |i| Address::parse(params, i))(input)?;
+            let primary_addr_param = addr.map(|a| IndexedParam {
+                param: Param::Addr(a),
+                payment_index: 0,
+            });
+
+            Ok((input, primary_addr_param))
+        }
+    }
+
+    /// The primary parser for <name>=<value> query-string
+    /// parameter pair.
+    pub fn zcashparam<'a, P: consensus::Parameters>(
+        params: &'a P,
+    ) -> impl Fn(&str) -> IResult<&str, IndexedParam> + 'a {
+        move |input| {
+            map_res(
+                separated_pair(indexed_name, char('='), recognize(qchars)),
+                move |r| to_indexed_param(params, r),
+            )(input)
+        }
+    }
+
+    /// Extension for the `alphanumeric0` parser which extends that parser with
+    /// the permitted characters identified in
+    fn alphanum_or(allowed: String) -> impl (Fn(&str) -> IResult<&str, &str>) {
+        move |input| {
+            input.split_at_position_complete(|item| {
+                let c = item.as_char();
+                !(c.is_alphanum() || allowed.contains(c))
+            })
+        }
+    }
+
+    pub fn qchars(input: &str) -> IResult<&str, &str> {
+        alphanum_or("-._~!$'()*+,;:@%".to_string())(input)
+    }
+
+    pub fn namechars(input: &str) -> IResult<&str, &str> {
+        alphanum_or("+-".to_string())(input)
+    }
+
+    pub fn indexed_name<'a>(input: &'a str) -> IResult<&'a str, (&'a str, Option<&'a str>)> {
+        let paramname = recognize(tuple((alpha1, namechars)));
+
+        tuple((
+            paramname,
+            opt(preceded(
+                char('.'),
+                recognize(tuple((one_of("123456789"), digit0))),
+            )),
+        ))(input)
+    }
+
+    pub fn parse_amount<'a>(input: &'a str) -> IResult<&'a str, Amount> {
+        map_res(
+            tuple((digit1, opt(preceded(char('.'), digit1)))),
+            |(whole_s, decimal_s): (&str, Option<&str>)| {
```

Comment:
Please add a test case where the whole part is greater than the maximum `i64`.

Reply:
https://github.com/zcash/librustzcash/pull/294/commits/32e85a65ea2733857a7d22df904d25a8798e4dab and the following two commits resolve this.